modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java

```diff
@@ -547,28 +549,163 @@ private <T> CompletableFuture<T> postEnlist(CompletableFuture<T> fut, boolean au
         }).thenCompose(x -> x);
     }
 
+    /**
+     * Evaluates the single-row request to the cluster for a read-only single-partition transaction.
+     *
+     * @param row Binary row.
+     * @param op Operation.
+     * @param <R> Result type.
+     * @return The future.
+     */
+    private <R> CompletableFuture<R> evaluateReadOnlyPrimaryNode(
+            BinaryRowEx row,
+            Function<ReplicationGroupId, ReplicaRequest> op
+    ) {
+        InternalTransaction tx = txManager.begin(observableTimestampTracker, true);
+
+        int partId = partitionId(row);
+
+        TablePartitionId tablePartitionId = new TablePartitionId(tableId, partId);
+
+        CompletableFuture<ReplicaMeta> primaryReplicaFuture = placementDriver.awaitPrimaryReplica(tablePartitionId, tx.startTimestamp())
+                .orTimeout(AWAIT_PRIMARY_REPLICA_TIMEOUT, TimeUnit.SECONDS);
+
+        CompletableFuture<R>  fut = primaryReplicaFuture.thenCompose(primaryReplica -> {
+            try {
+                ClusterNode node = clusterNodeResolver.apply(primaryReplica.getLeaseholder());
+
+                if (node == null) {
```

Comment:
How it's possible?

Reply:
Each time when we find a node in topology by its identifier it is possible.

Reply:
Yep. You are right.